### ISSUE TITLE
[BUG] get_pos_neg_edges neglects negative edges in ogbl datasets

### ISSUE THREAD

**Issue by Jn-Huang** [facebookresearch/SEAL_OGB#42]
## Bug Description

Thank you for reviewing this issue.

As titled, even when negative edges are given, e.g. `ogbl-ddi` or `ogbl-collab`, SEALDataset will resample negative edges regardlessly. See https://github.com/facebookresearch/SEAL_OGB/blob/21cda234ceb17fe757aa5d8ab598585a62336086/utils.py#L240

Since `ogbl-ddi` and `ogbl-collab` do not have  `edge_neg` in `split_edge['train']`, we will fail the `if` condition and resample negative edges in following lines. **This may not be a proper use for ogb dataset.**

## Error Reproduction

To justify this bug, we manually remove the negative edges in dataset and find that we can still run the code. It means that we are not using negative edges anyway.

After https://github.com/facebookresearch/SEAL_OGB/blob/21cda234ceb17fe757aa5d8ab598585a62336086/seal_link_pred.py#L431 add these lines:

```
print(f"{split_edge = }")

split_edge['valid']['edge_neg'] = None
split_edge['test']['edge_neg'] = None

print(f"{split_edge = }")
```

Then run

```
python seal_link_pred.py --dataset ogbl-ddi --num_hops 1 --ratio_per_hop 0.2 --use_edge_weight --eval_steps 1 --epochs 2 --runs 1  --dynamic_val --dynamic_test --train_percent 0.01 --val_percent 1 --test_percent 1
```

Outputs:
```
Results will be saved in results/ogbl-ddi_20230619211543
Command line input: python seal_link_pred.py --dataset ogbl-ddi --num_hops 1 --ratio_per_hop 0.2 --use_edge_weight --eval_steps 1 --epochs 2 --runs 1 --dynamic_val --dynamic_test --train_percent 0.01 --val_percent 1 --test_percent 1
 is saved.
split_edge = {'train': {'edge': tensor([[4039, 2424],
        [4039,  225],
        [4039, 3901],
        ...,
        [ 647,  708],
        [ 708,  338],
        [ 835, 3554]])}, 'valid': {'edge': tensor([[ 722,  548],
        [ 874, 3436],
        [ 838, 1587],
        ...,
        [3661, 3125],
        [3272, 3330],
        [1330,  776]]), 'edge_neg': tensor([[   0,   58],
        [   0,   84],
        [   0,   90],
        ...,
        [4162, 4180],
        [4168, 4260],
        [4180, 4221]])}, 'test': {'edge': tensor([[2198, 1172],
        [1205,  719],
        [1818, 2866],
        ...,
        [ 326, 1109],
        [ 911, 1250],
        [4127, 2480]]), 'edge_neg': tensor([[   0,    2],
        [   0,   16],
        [   0,   42],
        ...,
        [4168, 4259],
        [4208, 4245],
        [4245, 4259]])}}
split_edge = {'train': {'edge': tensor([[4039, 2424],
        [4039,  225],
        [4039, 3901],
        ...,
        [ 647,  708],
        [ 708,  338],
        [ 835, 3554]])}, 'valid': {'edge': tensor([[ 722,  548],
        [ 874, 3436],
        [ 838, 1587],
        ...,
        [3661, 3125],
        [3272, 3330],
        [1330,  776]]), 'edge_neg': None}, 'test': {'edge': tensor([[2198, 1172],
        [1205,  719],
        [1818, 2866],
        ...,
        [ 326, 1109],
        [ 911, 1250],
        [4127, 2480]]), 'edge_neg': None}}
Total number of parameters is 522946
SortPooling k is set to 244
100%|███████████████████████████████████| 7/7 [00:04<00:00,  1.51it/s]
100%|█████████████████████████████████| 84/84 [00:15<00:00,  5.46it/s]
100%|█████████████████████████████████| 84/84 [00:14<00:00,  5.73it/s]
Hits@20
Run: 01, Epoch: 01, Loss: 0.6864, Valid: 12.29%, Test: 7.35%
Hits@50
Run: 01, Epoch: 01, Loss: 0.6864, Valid: 20.69%, Test: 14.17%
Hits@100
Run: 01, Epoch: 01, Loss: 0.6864, Valid: 30.51%, Test: 22.49%
100%|███████████████████████████████████| 7/7 [00:02<00:00,  2.79it/s]
100%|█████████████████████████████████| 84/84 [00:14<00:00,  5.64it/s]
100%|█████████████████████████████████| 84/84 [00:14<00:00,  5.77it/s]
Hits@20
Run: 01, Epoch: 02, Loss: 0.6602, Valid: 32.83%, Test: 22.34%
Hits@50
Run: 01, Epoch: 02, Loss: 0.6602, Valid: 44.83%, Test: 36.21%
Hits@100
Run: 01, Epoch: 02, Loss: 0.6602, Valid: 53.67%, Test: 49.93%
Hits@20
Run 01:
Highest Valid: 32.83
Highest Eval Point: 2
   Final Test: 22.34
Hits@50
Run 01:
Highest Valid: 44.83
Highest Eval Point: 2
   Final Test: 36.21
Hits@100
Run 01:
Highest Valid: 53.67
Highest Eval Point: 2
   Final Test: 49.93
Hits@20
All runs:
Highest Valid: 32.83 ± nan
   Final Test: 22.34 ± nan
Hits@50
All runs:
Highest Valid: 44.83 ± nan
   Final Test: 36.21 ± nan
Hits@100
All runs:
Highest Valid: 53.67 ± nan
   Final Test: 49.93 ± nan
Total number of parameters is 522946
```

## Possible solution

Maybe we can change https://github.com/facebookresearch/SEAL_OGB/blob/21cda234ceb17fe757aa5d8ab598585a62336086/utils.py#L240 to 
```
if 'edge_neg' in split_edge[split]:
```
to fix this bug.

**Comment by muhanzhang** [comment]
@Jn-Huang Thanks! I think you are right. It is now fixed in 5874975. Previously, the code will ignore the provided negative edges in other OGB datasets and resample negative edges, whenever `'edge_neg'` is not present in `split_edge['train']`. Now, the code should individually check if `'edge_neg'` is in each split and selectively does the resampling. @jqmcginnis Could you verify that it does not affect the ogbl-vessel running? Thanks.

**Comment by jqmcginnis** [comment]
Good catch @Jn-Huang @muhanzhang  I checked both dynamic and static settings for `ogbl-vessel `(adding a print for debugging purposes to assess which route it's taking), and it's looking good to me!
